File: src/img.js
#!/usr/bin/env node

var chp = require('child_process')
var fs = require('fs');
var os = require('os')
var crypto = require('crypto');
var {Command} = require('commander');
var cheerio = require('cheerio');
var path = require('path')
var uuid = require('uuid')
var config = require('./config.js')
var pkg = require('../package.json');
var {
    requestRetry, 
    isDir, 
    safeMkdir,
    safeUnlink,
    sleep,
    optiImg,
} = require('./util.js')

process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';

function processDir(dir) {
    var files = fs.readdirSync(dir);
    for(var fname of files) {
        fname = path.join(dir, fname)
        processFile(fname)
    }
}

function processFile(fname) {
    if(!fname.endsWith('.html') &&
       !fname.endsWith('.md'))
        return;
    console.log('file: ' + fname);
    var content = fs.readFileSync(fname, 'utf-8');
    var imgs = new Map()
    if(fname.endsWith('.html'))
        content = processImg(content, imgs);
    else
        content = processImgMd(content, imgs);
    fs.writeFileSync(fname, content);
    var dir = path.join(path.dirname(fname), 'img')
    for(var [fname, img] of imgs.entries()) {
        fs.writeFileSync(path.join(dir, fname), img)
    }
}

function processImgMd(md, imgs, options={}) {
    options.imgPrefix = options.imgPrefix || 'img/'
    
    var re = /!\[.*?\]\((.*?)\)/g
    var rm;
    while(rm = re.exec(md)) {
        try {
            var url = rm[1]
            if(!url) continue
            if(options.pageUrl)
                url = new URL(url, options.pageUrl).toString()
            if(!url.startsWith('http')) continue
            url = encodeURI(url).replace(/%25/g, '%')
            
            var picname = crypto.createHash('md5').update(url).digest('hex') + ".png";
            console.log(`pic: ${url} => ${picname}`)
            
            if(!imgs.has(picname)) {
                var data = requestRetry('GET', url, {
                    headers: config.headers,
                    retry: config.retry,
                    timeout: config.timeout * 1000,
                }).body
                data = optiImg(data, config.optiMode, config.colors)
                imgs.set(picname, data);
                sleep(config.wait)
            }
            
            md = md.split(rm[1]).join(options.imgPrefix + picname)
        } catch(ex) {console.log(ex.toString())}
    }
    
    return md
}

function getImgSrc($img) {
	var url = ''
	for(var prop of config.imgSrc) {
		url = $img.attr(prop)
		if(url) break
	}
	return url
	
}

function processImg(html, imgs, options={}) {
    
    options.imgPrefix = options.imgPrefix || 'img/'
    
    var $ = cheerio.load(html);
    
    var $imgs = $('img');

    for(var i = 0; i < $imgs.length; i++) {
        try {
            var $img = $imgs.eq(i);
            var url = getImgSrc($img)
            if(!url) continue
            if(!url.startsWith('http')) {
                if(options.pageUrl)
                    url = new URL(url, options.pageUrl).toString()
                else
                    continue
            }
            url = encodeURI(url).replace(/%25/g, '%')
            
            var picname = crypto.createHash('md5').update(url).digest('hex') + ".png";
            console.log(`pic: ${url} => ${picname}`)
            
            if(!imgs.has(picname)) {
                var data = requestRetry('GET', url, {
                    headers: config.headers,
                    retry: config.retry,
                    timeout: config.timeout * 1000,
                }).body
                data = optiImg(data, config.optiMode, config.colors)
                imgs.set(picname, data);
                sleep(config.wait)
            }
            
            $img.attr('src', options.imgPrefix + picname);
        } catch(ex) {console.log(ex.toString())}
    }
    
    return $.html();
}

function main() {
    
    var cmder = new Command()
    cmder.name('crawl-img')
         .version(pkg.version)
         .arguments('<fname>', 'img file name of dir name')
         .option('-r, --retry <retry>', 'time of retrying', parseInt, 10)
         .option('-w, --wait <wait>', 'wait time in seconds', parseFloat, 0)
         .option('-m, --opti-mode <mode>', 'image optimization mode', 'quant')
         .option('-c, --colors <colors>', 'num of colors for image optimization mode', parseInt, 8)
         .option('-t, --timeout <timeout>', 'timeout', parseInt, 8)
    args = cmder.parse()
    
    config.retry = args.retry
    config.wait = args.wait
    config.optiMode = args.optiMode
    config.colors = args.colors
    config.timeout = args.timeout
    
    var fname = args.args[0]
    if(!fs.existsSync(fname))
    {
        console.log('file or dir not found');
        return
    }
    
    if(isDir(fname)) {
        safeMkdir(path.join(fname, 'img'))
        processDir(fname)
    } else {
        safeMkdir(path.join(path.dirname(fname), 'img'))
        processFile(fname)
    }
}

module.exports = processImg

if(module == require.main) main();

File: EpubCrawler/img.py
# -*- coding: utf-8 -*-

from concurrent.futures import ThreadPoolExecutor
from urllib.parse import urljoin, quote_plus
import hashlib
from pyquery import PyQuery as pq
import time
import re
import base64
from .util import *
from .config import config

img_pool = ThreadPoolExecutor(5)
RE_DATA_URL = r'^data:image/\w+;base64,'


def set_img_pool(pool):
    global img_pool
    img_pool = pool
    
def get_img_src(el_img):
    url = ''
    for prop in config['imgSrc']:
        url = el_img.attr(prop)
        if url: break
    return url
    
def tr_download_img_safe(url, imgs, picname):
    try:
        tr_download_img(url, imgs, picname)
    except Exception as ex:
        print(f'{url} 下载失败：{ex}')

def tr_download_img(url, imgs, picname):
    hash = hashlib.md5(url.encode('utf-8')).hexdigest()
    cache = load_img(hash, config['optiMode'])
    if cache is not None and config['cache']:
        print(f'{url} 已存在于本地缓存中')
        imgs[picname] = cache
        return

    data = request_retry(
        'GET', url,
        headers=config['headers'],
        check_status=config['checkStatus'],
        retry=config['retry'],
        timeout=config['timeout'],
        proxies=config['proxy'],
    ).content
    print(f'{url} 下载成功')
    data = opti_img(data, config['optiMode'], config['colors']) or b''
    imgs[picname] = data
    save_img(hash, config['optiMode'], data)
    time.sleep(config['wait'])
    
def process_img_data_url(url, el_img, imgs, **kw):
    if not re.search(RE_DATA_URL, url):
        return False
    picname = hashlib.md5(url.encode('utf-8')).hexdigest() + '.png'
    print(f'pic: {url} => {picname}')
    if picname not in imgs:
        enco_data = re.sub(RE_DATA_URL, '', url)
        data = base64.b64decode(enco_data.encode('utf-8'))
        data = opti_img(data, config['optiMode'], config['colors'])
        imgs[picname] = data
    el_img.attr('src', kw['img_prefix'] + picname)
    return True
    
def process_img(html, imgs, **kw):
    kw.setdefault('img_prefix', 'img/')
    
    root = pq(html)
    el_imgs = root('img')
    hdls = []
    
    for i in range(len(el_imgs)):
        el_img = el_imgs.eq(i)
        url = get_img_src(el_img)
        if not url: continue
        if process_img_data_url(url, el_img, imgs, **kw):
            continue
        if not url.startswith('http'):
            if kw.get('page_url'):
                url = urljoin(kw.get('page_url'), url)
            else: continue
        
        picname = hashlib.md5(url.encode('utf-8')).hexdigest() + '.png'
        print(f'pic: {url} => {picname}')
        if picname not in imgs:
            hdl = img_pool.submit(tr_download_img_safe, url, imgs, picname)
            hdls.append(hdl)
            
        el_img.attr('src', kw['img_prefix'] + picname)
        el_img.attr('data-original-src', url)
        
    for h in hdls: h.result()
    return root.html()
    
    